packages/v2/contracts/ANTv2.sol

```diff
@@ -0,0 +1,171 @@
+
+    bytes32 public DOMAIN_SEPARATOR;
+
+    // bytes32 public constant PERMIT_TYPEHASH = keccak256("Permit(address owner,address spender,uint256 value,uint256 nonce,uint256 deadline)");
+    bytes32 public constant PERMIT_TYPEHASH = 0x6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9;
+    mapping (address => uint256) public nonces;
+
+    // bytes32 public constant TRANSFER_WITH_AUTHORIZATION_TYPEHASH =
+    //     keccak256("TransferWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)");
+    bytes32 public constant TRANSFER_WITH_AUTHORIZATION_TYPEHASH = 0x7c7c6cdb67a18743f49ec6fa9b35f50d52ed05cbed4cc592e13b44501c1a2267;
+    mapping (address => mapping (bytes32 => bool)) public authorizationState;
+
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+    event Transfer(address indexed from, address indexed to, uint256 value);
+    event AuthorizationUsed(address indexed authorizer, bytes32 indexed nonce);
+    event ChangeMinter(address indexed minter);
+
+    modifier onlyMinter {
+        require(msg.sender == minter, "ANTV2:NOT_MINTER");
+        _;
+    }
+
+    constructor(uint256 chainId, address initialMinter) public {
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                keccak256('EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)'),
+                keccak256(bytes(name)),
+                keccak256(bytes('1')),
+                chainId,
+                address(this)
+            )
+        );
+
+        _changeMinter(initialMinter);
+    }
+
+    function _validateSignedData(address signer, bytes32 encodeData, uint8 v, bytes32 r, bytes32 s) internal view {
+        bytes32 digest = keccak256(
+            abi.encodePacked(
+                "\x19\x01",
+                DOMAIN_SEPARATOR,
+                encodeData
+            )
+        );
+        address recoveredAddress = ecrecover(digest, v, r, s);
+        // Explicitly disallow authorizations for address(0) as ecrecover returns address(0) on malformed messages
+        require(recoveredAddress != address(0) && recoveredAddress == signer, "ANTV2:INVALID_SIGNATURE");
+    }
+
+    function _changeMinter(address newMinter) internal {
+        minter = newMinter;
+        emit ChangeMinter(newMinter);
+    }
+
+    function _mint(address to, uint256 value) internal {
+        totalSupply = totalSupply.add(value);
+        balanceOf[to] = balanceOf[to].add(value);
+        emit Transfer(address(0), to, value);
+    }
+
+    function _burn(address from, uint value) internal {
+        // Balance is implicitly checked with SafeMath's underflow protection
+        balanceOf[from] = balanceOf[from].sub(value);
+        totalSupply = totalSupply.sub(value);
+        emit Transfer(from, address(0), value);
+    }
+
+    function _approve(address owner, address spender, uint256 value) private {
```

Comment:
Not against, but we are not providing any built-in solution to avoid front-running issues here

Reply:
Yes... the EIP now actually states that the contract should not enforce anything (as this will break contract-to-contract interactions), and that frontends should handle this.